From: Dmitry Borodaenko <angdraug@debian.org>
Date: Mon, 12 Oct 2020 11:11:26 -0700
Subject: Implement has_option in Xsession

Now that x11-common moved this function from 20x11-common_process-args
to Xsession (see #920778), gdm's version of Xsession also has to
implement it.

https://bugs.debian.org/972108

Modified by: Illia Bobyr <illia.bobyr@gmail.com>

A subsequent patch `debian/run_xsession.d.patch` adds code that sources
`/etc/X11/Xsession.d/`.  Scripts in this folder expect `has_option` to be
defined.  So we need `OPTIONS` and `has_option` to be located before the code
that will source `Xsession.d`.

Forwarded: not-needed
---
 data/Xsession.in | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/data/Xsession.in b/data/Xsession.in
index 2e4de4f..72cef98 100755
--- a/data/Xsession.in
+++ b/data/Xsession.in
@@ -189,6 +189,24 @@ if [ -d /etc/X11/xinit/xinitrc.d ]; then
     done
 fi
 
+OPTIONS="$(
+  if [ -r "$OPTIONFILE" ]; then
+    cat "$OPTIONFILE"
+  fi
+  if [ -d /etc/X11/Xsession.options.d ]; then
+    run-parts --list --regex '\.conf$' /etc/X11/Xsession.options.d
+  fi
+)"
+
+has_option() {
+  # Ensure that a later no-foo overrides an earlier foo
+  if [ "$(echo "$OPTIONS" | grep -Eo "^(no-)?$1\>" | tail -n 1)" = "$1" ]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
 if [ "x$command" = "xdefault" ] ; then
   if [ -x "$HOME/.Xclients" ]; then
     command="$HOME/.Xclients"
